### ISSUE TITLE
Not using DHCP resolver

### ISSUE THREAD

**Issue by blacoste** [siacs/Conversations#3383]
[logcat.txt](https://github.com/siacs/Conversations/files/2874834/logcat.txt)
#### General information

* **Version:** 2.3.12+fcr
* **Device:** eg OnePlus 6
* **Android Version:** eg Android 9 OxygenOS 9.0.4
* **Server software:** prosody 0.9
* **Conversations source:** F-Droid

#### Steps to reproduce

Connection fails when connecting to a server on same private network and when public DNS target != Private DNS target.

Private DNS records from server provided by DHCP:
```
_xmpp-client._tcp.<domain> 86400 IN   SRV     10 1 5222 <sub>.<domain>
<sub>.<domain>   0       IN      A       192.168.8.20
```
Public DNS records
```
_xmpp-client._tcp.<domain> 86400 IN   SRV     10 1 5222 <sub>.<domain>
<sub>.<domain>   60       IN      A       90.XXX.XXX.XXX
```
#### Expected result

Should connect. Does not connect anymore (server not found) whenever I connect to home wifi.

#### Actual result

Conversation resolves server address with public DNS server while when I try while 
`nslookup <sub>.<domain> `
on a tty, it is a private address that is returned (from private DNS provided by DHCP).
Does the application uses a hardcoded specific resolver?
This is working for other app like Connectbot (SSH client), using same `<sub>.<domain>` address

#### Debug output

see file attached.

Note if it may help:
Was working for years. Recent updates : Handcet ROM, Conversations app.

**Comment by licaon-kter** [comment]
What does a query from Termux look like?

**Comment by blacoste** [comment]
Termux resolves to private address using "su", then "nslookup".
Conversation logcat resolves to public address.
No IPV6 records for this domain.

**Comment by blacoste** [comment]
Also tested with an old Honor 7 (Android 6.0) and it works.
Issue seems happens on Android 9 OxygenOS 9.0.4

**Comment by blacoste** [comment]
Makes me think this could be a bug introduced by the ROM about new Android feature "Private DNS".

getprop | grep dns
[net.dns1]: [192.168.<x>.<x>]
[net.dns2]: [8.8.8.8]

**Comment by iNPUTmice** [comment]
This is the class where Conversations figures out what DNS server to use: https://github.com/siacs/Conversations/blob/master/src/main/java/eu/siacs/conversations/utils/AndroidUsingLinkProperties.java

If you want to debug this you can start by adding something like `Log.d(Config.LOGTAG,"DNS servers"+servers);` to line [62](https://github.com/siacs/Conversations/blob/master/src/main/java/eu/siacs/conversations/utils/AndroidUsingLinkProperties.java#L62) to see what DNS servers are reported by the system.
The code as you can see in the lines above that will give preference to 'active networks'. It might also be as simple as a cache issue. Does the problem go away if you restart Conversations (kill app) when wifi is already active?

**Comment by blacoste** [comment]
Killing the app on wifi network does not solve the issue. 
Since OS upgrade, I have never reached the server through wifi (including with reboot / reboot with Airplane mode + wifi).
I will log the DNS used when possible.

**Comment by pushytoxin** [comment]
@blacoste you should fix your issue by configuring your router to forward TCP/5222 to the XMPP server from every interface, not just the external one.

Having an IM nexus behind NAT is surely a blast. Get ready for days of debugging once you want to setup VOIP.

**Comment by blacoste** [comment]
@pushytoxin of course that would be the easiest solution if my internet provider (Orange) would provide a fully functional router with Loopback forwarding of public IP. But that's not the case, probably because they think only professionals would pay for that. That's why using [DNS "hack"](https://martin-jones.com/2013/12/10/orange-livebox-2-0-port-forwarding-and-nat-loopback/) is the easiest solution I found . 

For VOIP, it is OK with STUN/TURN severs.

Back to the main topic, I got some news. If DHCP is only providing one DNS server, the other one is automatically filled with 8.8.8.8. (verified with another OnePlus 6 on another network, with and without secondary DNS sever). Aside that sharing DNS requests to google is a ****** privacy violation  (legal or not is not the matter), it seems that it can't be disabled.

So I think we can close this issue and I will look after answears from Android or ROM specific implementation.
